File: CHUONG_10/10.2_Singleton/app/src/main/java/com/example/myapplication/ui/StudentAdapter.java
package com.example.myapplication.ui;

import android.text.Layout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.data.model.Student;
import com.example.myapplication.databinding.ItemStudentBinding;
import com.example.myapplication.utils.AppUtils;

import java.util.ArrayList;
import java.util.List;

public class StudentAdapter extends RecyclerView.Adapter<StudentAdapter.StudentViewHolder>{
    private final List<Student> mStudents = new ArrayList<>();
    private final OnClickListener mListener;

    public StudentAdapter(OnClickListener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public StudentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        ItemStudentBinding binding = ItemStudentBinding.inflate(layoutInflater, parent, false);
        return new StudentViewHolder(binding, mListener);
    }

    @Override
    public void onBindViewHolder(@NonNull StudentViewHolder holder, int position) {
        holder.bind(mStudents.get(position));
    }

    @Override
    public int getItemCount() {
        return mStudents.size();
    }

    public void updateStudents(List<Student> students) {
        int oldSize = mStudents.size();
        mStudents.clear();
        mStudents.addAll(students);
        notifyItemRangeRemoved(0, oldSize);
        notifyItemRangeInserted(0, students.size());
    }

    public interface OnClickListener {
        void onClick(Student student);
    }

    public static class StudentViewHolder extends RecyclerView.ViewHolder {
        private final ItemStudentBinding mBinding;
        private final OnClickListener mListener;
        public StudentViewHolder(ItemStudentBinding binding, OnClickListener listener) {
            super(binding.getRoot());
            mBinding = binding;
            mListener = listener;
        }
        public void bind(Student student){
            mBinding.textStudentId.setText(student.getmId());
            mBinding.textFullName.setText(student.getmFullName().toString());
            mBinding.textGpa.setText(student.getmGender());
            mBinding.imageAvatar.setImageResource(AppUtils.getAvatar(student.getmGender()));
            mBinding.getRoot().setOnClickListener(v -> mListener.onClick(student));
        }
    }
}

File: CHUONG_5/5.11_RecyclerView_Kotlin/app/src/main/java/com/example/myapplication/StudentAdapter.kt
package com.example.myapplication

import android.view.LayoutInflater
import android.view.OrientationEventListener
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import org.w3c.dom.Text

class StudentAdapter(
    private val listener: OnStudentClickListener
): RecyclerView.Adapter<StudentAdapter.StudentViewHolder>() {
    private val students = mutableListOf<Student>()
    class StudentViewHolder(
        itemView: View,
        private val listener: OnStudentClickListener
    ): ViewHolder(itemView) {
        private val imageAvatar: ImageView = itemView.findViewById(R.id.image_avatar)
        private val textId: TextView = itemView.findViewById(R.id.text_student_id)
        private val textFullName: TextView = itemView.findViewById(R.id.text_full_name)
        private val textGpa: TextView = itemView.findViewById(R.id.text_gpa)

        fun bind(student: Student){
            textFullName.text = student.fullName.toString()
            textGpa.text = student.gpa.toString()
            textId.text = student.id
            imageAvatar.setImageResource(Utils.getAvatar(student.gender))
            itemView.setOnClickListener{
                listener.onClick(student)
            }
        }
    }

    fun updateStudents(studentList: List<Student>){
        val oldSize = students.size
        students.clear()
        students.addAll(studentList)
        notifyItemRangeRemoved(0, oldSize)
        notifyItemRangeInserted(0, students.size)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): StudentViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_student, parent, false)
        return StudentViewHolder(view, listener)
    }

    override fun onBindViewHolder(holder: StudentAdapter.StudentViewHolder, position: Int) {
       val student = students[position]
        holder.bind(student)
    }

    override fun getItemCount(): Int {
        return students.size
    }

    interface OnStudentClickListener {
        fun onClick(student: Student)
    }

}